src/waggle/data/vision.py

```diff
@@ -121,12 +172,20 @@ def stream(self):
         with self.capture:
             yield from self.capture.stream()
 
+    def record(self, duration, file_path="./sample.mp4", skip_second=1):
+        return self.capture.record(duration, file_path, skip_second)
+
 
 class _Capture:
     def __init__(self, device, format):
         self.device = device
         self.format = format
         self.context_depth = 0
+        self.enable_daemon = False
+        self.daemon_need_to_stop = False
+        self._ready_for_next_frame = Queue(maxsize=1)
+        self.daemon = threading.Thread(target=self._run)
```

Comment:
You probably want to set `Thread(..., daemon=True)`.

If not, you main thread will wait on this forever to finish instead of exiting.

Reply:
Just a coding / technical question as I don't know. I have `.join()` for the child thread when the main thread exits. Would making the child thread a daemon matter? Should I remove the `.join()` such that the main thread can exit no matter how the child thread runs?

```python
def __exit__(self, exc_type, exc_val, exc_tb):
    self.context_depth -= 1
    if self.context_depth == 0:
        if self.enable_daemon:
            self.daemon_need_to_stop = True
            self.daemon.join()
        self.capture.release()
```

Reply:
Yeah, I'd remove the join. Ideally, the "needs to stop" event allows us to _attempt_ to gracefully stop the background thread but I don't think we should require it. The OS will clean things up in that case.

Reply:
For thread safety, you probably want to make `self.daemon_need_to_stop` an Event too, instead of just a variable.

Reply:
Thanks for the comment. Just wanted to confirm what I think is correct. Pushed another commit to reflect this.